src/software/amazon/ion/Timestamp.java

```diff
@@ -690,7 +705,20 @@ private Timestamp(BigDecimal millis, Integer localOffset)
     {
         if (millis == null) throw new NullPointerException("millis is null");
 
-        long ms = millis.longValue();
+        // check bounds to avoid hanging when calling longValue() on decimals with large positive exponents,
+        // e.g. 1e10000000
+        if(millis.compareTo(MINIMUM_ALLOWED_FRACTIONAL_MILLIS) < 0 ||
+            FRACTIONAL_MILLIS_UPPER_BOUND.compareTo(millis) < 0) {
+            throw new IllegalArgumentException("millis: " + millis + " is outside of valid the range: from "
+                + MINIMUM_ALLOWED_FRACTIONAL_MILLIS
+                + " (0001T)"
+                + ", inclusive, to "
+                + FRACTIONAL_MILLIS_UPPER_BOUND
+                + " (10000T)"
+                + " , exclusive");
+        }
+
+        long ms = millis.divideToIntegralValue(BigDecimal.ONE).longValue();
```

Comment:
Divide-by-one is, IMO, rather subtle.  I think it's more clear to address things head-on. The idiom I [recommended for Jackson](https://github.com/FasterXML/jackson-modules-java8/pull/84/files#r223886551) applies here:

```
long result;
if (value.signum() == 0 || value.scale() < -63 || value.precision() - value.scale() <= 0) {
    // Quickly handle 0 ||  no low-order bits  || no integer part, only a fraction
    result = 0;
} else {
    result = value.longValue();
}
```

This could replace all of the above, since the `set_fields_from_millis` will check the bounds.